### PR TITLE
Add support for XPath scalar functions - xpath,xpath_string,xpath_boolean,xpath_double

### LINKED ISSUE #3888: Support for XPath functions
Both Spark and Hive support XPath functions:

https://cwiki.apache.org/confluence/display/Hive/LanguageManual+XPathUDF
https://github.com/apache/hive/tree/master/ql/src/java/org/apache/hadoop/hive/ql/udf/xml

https://spark.apache.org/docs/2.3.0/api/sql/index.html
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/xml/UDFXPathUtil.java
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/xml/xpath.scala

### PR DESCRIPTION
**Pull request by sajjoseph**
Supersedes #4024 Fixes #3888

Added support for the following scalar functions. 

xpath - returns an array of strings that matched the criteria
xpath_string - returns a string that matched the criteria
xpath_boolean - returns the boolean value based on the xpath expression
xpath_double - returns a double-precision value after evaluating the xpath expression

Example:
```
trino> select xpath ('<a><b id="1"><c/></b><b id="2"><c/></b></a>','/descendant::c/ancestor::b/@id');
 _col0
--------
 [1, 2]
 
 trino> SELECT xpath_string ('<a><b>bb</b><c>cc</c></a>', 'a/b');
 _col0
-------
 bb


trino> SELECT xpath_double('<a><b class="odd">1</b><b class="even">2</b><b class="odd">4</b><c>8</c></a>', 'sum(a/*)') ;
 _col0
-------
  15.0
```
 
Few more sample queries can be found here - https://cwiki.apache.org/confluence/display/Hive/LanguageManual+XPathUDF

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/operator/scalar/TestXPathFunction.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        assertFunction("xpath_string('<a><b>b1</b><b id=\"b_2\">b2</b></a>','a/b[@id=\"b_2\"]')", VarcharType.createVarcharType(34), "b2");
+        assertInvalidFunction("xpath_string('\"\"','detail/reason')", INVALID_FUNCTION_ARGUMENT, "Error parsing xml data '\"\"'");
+    }
+
+    @Test
+    public void testXpath()
+    {
+        assertFunction("xpath('<a><b id=\"foo\">b1</b><b id=\"bar\">b2</b></a>','//@id')",
+                new ArrayType(VARCHAR),
+                Arrays.asList("foo", "bar"));
```
Comment (ebyhr) [inline core/trino-main/src/test/java/io/trino/operator/scalar/TestXPathFunction.java:47]: Use `ImmutableList.of()`.

File: core/trino-main/src/main/java/io/trino/operator/scalar/XPathFunctions.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        }
+        return utf8Slice(s);
+    }
+
+    @SqlNullable
+    @Description("Returns a string array of values within xml nodes that match the xpath expression")
+    @ScalarFunction("xpath")
+    @LiteralParameters({"x", "y"})
+    @SqlType("array(varchar)")
+    public static Block xpath(@SqlType("varchar(y)") Slice xml, @SqlType("varchar(x)") Slice path)
```
Comment (ebyhr) [inline core/trino-main/src/main/java/io/trino/operator/scalar/XPathFunctions.java:60]: Why `x` and `y` are flipped? Subsequent functions also.

**Review by ebyhr — COMMENTED** [review]
Left initial comments. I'm looking into other database behavior. 

Please add the documentation. 

The below comment about the argument order isn't yet resolved, so let me remind here. 
https://github.com/trinodb/trino/pull/4024#discussion_r442387326

**Comment by wgzhao** [comment]
@sajjoseph Are you still continuing the development of this feature? I need this feature in my work and I migrated the related functions in Hive using the same approach, can you continue to develop this PR and have it merged into the master branch

**Review by tooptoop4 — APPROVED** [review]

**Comment by mosabua** [comment]
:wave: @sajjoseph  - this PR has become inactive. If you're still interested in working on it, please let us know. @ebyhr might have some insight how/if to proceed.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Comment by tooptoop4** [comment]
this lit 🔥

**Comment by colebow** [comment]
Closing this one out due to inactivity, but please reopen if you would like to pick this back up. It seems like there's a decent amount of interest here, so if anyone else would like to take this on and get it updated, go for it.

**Pull request closed without merging**